torch/distributed/fsdp/_unshard_param_utils.py

```diff
@@ -216,6 +230,93 @@ def _unshard_params(
                     handle._training_state = HandleTrainingState.IDLE
 
 
+@contextlib.contextmanager
+def _unshard_params_recurse(
+    module: nn.Module,
+    state: _FSDPState,
+    recurse: bool,
+    writeback: bool,
+    rank0_only: bool,
+    offload_to_cpu: bool,
+    with_grads: bool,
+):
+    """
+    This is a helper for :func:`_unshard_params` that recursively calls
+    :func:`_unshard_fsdp_state_params` on FSDP states if ``recurse=True``.
+    """
+    _validate_unshard_params_args(
+        state, writeback, rank0_only, offload_to_cpu, with_grads
+    )
+    if recurse:
+        with contextlib.ExitStack() as stack:
+            # TODO (awgu): The traversal function does not traverse through
+            # incompatible composable APIs. Verify if this is the desired
```

Comment:
Could you elaborate, what's an example of this?

Reply:
```
fully_shard(
  Module(
    replicate(
      Submodule(
        fully_shard(Subsubmodule),
        Subsubmodule,
    ),
    Submodule,
)
```
Because the traversal utils do not go through incompatible composable APIs (here, `replicate`), calling `_unshard_params` on the root `Module` will not unshard the parameters of the fully sharded `Subsubmodule`.